transport/ProbeServer.cpp

```diff
@@ -0,0 +1,224 @@
+void ProbeServer::onUnregistered(Endpoint& endpoint)
+{
+    const std::lock_guard<std::mutex> locker(_endpointsLock);
+
+    if (endpoint.getTransportType() == ice::TransportType::UDP)
+    {
+        auto port = endpoint.getLocalPort().getPort();
+        _udpPorts.erase(port);
+
+        std::remove_if(_probeCandidates.begin(), _probeCandidates.end(), [port](const transport::ProbeCandidate& c) {
+            return c.transport == ice::TransportType::UDP && c.address.getPort() == port;
+        });
+    }
+}
+
+// ServerEndpoint::IEvents
+void ProbeServer::onServerPortRegistered(ServerEndpoint& endpoint)
+{
+    const std::lock_guard<std::mutex> locker(_endpointsLock);
+
+    auto port = endpoint.getLocalPort().getPort();
+    auto it = _tcpPorts.find(port);
+    if (it != _tcpPorts.end())
+    {
+        return;
+    }
+
+    if (!_iceConfig.publicIpv4.empty() && endpoint.getLocalPort().getFamily() == AF_INET)
+    {
+        auto address = _iceConfig.publicIpv4;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ProbeCandidate{ice::TransportType::TCP, address});
+
+        _tcpPorts.insert(port);
+
+        logger::debug("Registered TCP server endpoint %s:%u", _name, address.ipToString().c_str(), port);
+    }
+
+    if (!_iceConfig.publicIpv6.empty() && endpoint.getLocalPort().getFamily() == AF_INET6)
+    {
+        auto address = _iceConfig.publicIpv6;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ProbeCandidate{ice::TransportType::TCP, address});
+
+        _tcpPorts.insert(port);
+
+        logger::debug("Registered TCP server endpoint %s:%u", _name, address.ipToString().c_str(), port);
+    }
+}
+
+void ProbeServer::onServerPortUnregistered(ServerEndpoint& endpoint)
+{
+    // Re-register to keep listening on probes;
+    // TcpServerEndpoint unregisters the listener after dispatching onIceTcpConnect
+    endpoint.registerListener(_probeCredentials.first, this);
+}
+
+void ProbeServer::onIceTcpConnect(std::shared_ptr<Endpoint> endpoint,
+    const SocketAddress& source,
+    const SocketAddress& target,
+    memory::UniquePacket packet)
+{
+    if (endpoint->getTransportType() == ice::TransportType::TCP)
+    {
+        replyStunOk(*endpoint, source, std::move(packet));
+
+        // wait for while to ensure the reply is sent,
+        // before shutting down endpoint
+        u_int32_t gracePeriodMs = 1000;
+        usleep(gracePeriodMs * 1000);
+
+        endpoint->stop(this);
+    }
+}
+
+// Endpoint::IStopEvents
+void ProbeServer::onEndpointStopped(Endpoint* endpoint) {}
+
+void ProbeServer::replyStunOk(Endpoint& endpoint, const SocketAddress& destination, memory::UniquePacket packet)
+{
+    uint64_t timestamp = utils::Time::getAbsoluteTime();
+    const void* data = packet->get();
+
+    auto* stunMessage = ice::StunMessage::fromPtr(data);
+
+    if (stunMessage && stunMessage->isValid() && stunMessage->header.isRequest() &&
+        stunMessage->isAuthentic(_probeCredentials.second))
```

Comment:
Not sure if this verification fails and is a TCP we want to close sockets as soon as possible

Reply:
I will try to add it.

Reply:
Fixed.